slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java

```diff
@@ -0,0 +1,127 @@
+      final LowWatermarkDao lowWatermarkDao) {
+    this.jdbi = jdbi;
+    this.signedBlocksDao = signedBlocksDao;
+    this.signedAttestationsDao = signedAttestationsDao;
+    this.lowWatermarkDao = lowWatermarkDao;
+  }
+
+  public void pruneForValidator(
+      final int validatorId, final long epochsToKeep, final long slotsPerEpoch) {
+    checkArgument(
+        epochsToKeep > 0,
+        "epochsToKeep to keep must be a positive value, but was %s",
+        epochsToKeep);
+    checkArgument(
+        slotsPerEpoch > 0, "slotsPerEpoch must be a positive value, but was %", slotsPerEpoch);
+    final long slotsToKeep = Math.max(epochsToKeep / slotsPerEpoch, 1);
+    pruneBlocks(validatorId, slotsToKeep);
+    pruneAttestations(validatorId, epochsToKeep);
+  }
+
+  private void pruneBlocks(final int validatorId, final long slotsToKeep) {
+    final Optional<UInt64> pruningSlot =
+        jdbi.inTransaction(
+            READ_UNCOMMITTED,
+            h -> {
+              lockForValidator(h, LockType.BLOCK, validatorId);
+              final Optional<UInt64> watermarkSlot =
+                  lowWatermarkDao
+                      .findLowWatermarkForValidator(h, validatorId)
+                      .map(SigningWatermark::getSlot);
+              final Optional<UInt64> slot = signedBlocksDao.findMaxSlot(h, validatorId);
+              final Optional<UInt64> slotToPruneTo =
+                  calculatePruningMark(slotsToKeep, slot, watermarkSlot);
+              slotToPruneTo.ifPresent(
+                  s -> lowWatermarkDao.updateSlotWatermarkFor(h, validatorId, s));
+              return slotToPruneTo;
+            });
+
+    pruningSlot.ifPresent(
+        s ->
+            jdbi.useTransaction(
+                READ_UNCOMMITTED, h -> signedBlocksDao.deleteBlocksBelowSlot(h, validatorId, s)));
+  }
+
+  private void pruneAttestations(final int validatorId, final long epochsToKeep) {
+    final Optional<UInt64> pruningEpoch =
+        jdbi.inTransaction(
+            READ_UNCOMMITTED,
+            h -> {
+              lockForValidator(h, LockType.ATTESTATION, validatorId);
+              final Optional<UInt64> watermarkEpoch =
+                  lowWatermarkDao
+                      .findLowWatermarkForValidator(h, validatorId)
+                      .map(SigningWatermark::getTargetEpoch);
+              final Optional<UInt64> epoch =
+                  signedAttestationsDao.findMaxTargetEpoch(h, validatorId);
+              final Optional<UInt64> epochToPruneTo =
+                  calculatePruningMark(epochsToKeep, epoch, watermarkEpoch);
+              epochToPruneTo.ifPresent(
+                  e -> lowWatermarkDao.updateEpochWatermarksFor(h, validatorId, e, e));
+              return epochToPruneTo;
+            });
+
+    pruningEpoch.ifPresent(
+        e ->
+            jdbi.useTransaction(
+                READ_UNCOMMITTED,
+                h -> signedAttestationsDao.deleteAttestationsBelowEpoch(h, validatorId, e)));
```

Comment:
is it worth calling this "deleteAttestationsbelowWatermark"? no need to pass "e" in - and thus the signedAttestationsDao function becomes a bit safer... (but then has dependency on watermark :( )

Reply:
I like this, i'll update the DAO to do this. done